File: football-api/app/api/v1/routers/mcp_tokens.py
import uuid
from datetime import datetime, timedelta, timezone
from uuid import uuid4

from fastapi import APIRouter
from sqlalchemy import select, update

from app.core.dependencies import DB, CurrentPlayer
from app.core.exceptions import ForbiddenError, NotFoundError
from app.models.mcp_token import MCPToken
from app.schemas.mcp_token import ExpiresIn, MCPTokenCreate, MCPTokenCreated, MCPTokenResponse

router = APIRouter(prefix="/mcp-tokens", tags=["mcp-tokens"])


def _compute_expires_at(expires_in: ExpiresIn | None) -> datetime | None:
    if expires_in == ExpiresIn.h24:
        return datetime.now(tz=timezone.utc) + timedelta(hours=24)
    if expires_in == ExpiresIn.d7:
        return datetime.now(tz=timezone.utc) + timedelta(days=7)
    return None


@router.post("", response_model=MCPTokenCreated, status_code=201)
async def create_token(body: MCPTokenCreate, db: DB, current_player: CurrentPlayer):
    raw, hashed, prefix = MCPToken.generate()
    expires_at = _compute_expires_at(body.expires_in)
    token_id = uuid4()
    now = datetime.now(tz=timezone.utc)

    token = MCPToken(
        id=token_id,
        player_id=current_player.id,
        name=body.name,
        token_hash=hashed,
        token_prefix=prefix,
        expires_at=expires_at,
        created_at=now,
    )
    db.add(token)
    await db.flush()

    return MCPTokenCreated(
        id=token_id,
        name=body.name,
        token=raw,
        token_prefix=prefix,
        expires_at=expires_at,
        created_at=now,
    )


@router.get("", response_model=list[MCPTokenResponse])
async def list_tokens(db: DB, current_player: CurrentPlayer):
    result = await db.execute(
        select(MCPToken)
        .where(MCPToken.player_id == current_player.id, MCPToken.revoked_at.is_(None))
        .order_by(MCPToken.created_at.desc())
    )
    tokens = result.scalars().all()
    now = datetime.now(tz=timezone.utc)

    return [
        MCPTokenResponse(
            id=t.id,
            name=t.name,
            token_prefix=t.token_prefix,
            expires_at=t.expires_at,
            created_at=t.created_at,
            last_used_at=t.last_used_at,
            is_expired=t.expires_at is not None and t.expires_at < now,
        )
        for t in tokens
    ]


@router.delete("/{token_id}", status_code=204)
async def revoke_token(token_id: uuid.UUID, db: DB, current_player: CurrentPlayer):
    result = await db.execute(
        select(MCPToken).where(MCPToken.id == token_id, MCPToken.revoked_at.is_(None))
    )
    token = result.scalar_one_or_none()

    if not token:
        raise NotFoundError("Token not found")
    if token.player_id != current_player.id:
        raise ForbiddenError("Cannot revoke another player's token")

    await db.execute(
        update(MCPToken)
        .where(MCPToken.id == token_id)
        .values(revoked_at=datetime.now(tz=timezone.utc))
    )

File: football-api-go/internal/handlers/mcp_tokens.go
package handlers

import (
	"context"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/thiagotn/football-manager/football-api-go/internal/apierror"
	"github.com/thiagotn/football-manager/football-api-go/internal/db"
	"github.com/thiagotn/football-manager/football-api-go/internal/middleware"
)

type MCPTokenStore interface {
	GenerateMCPToken() (raw, hash, prefix string, err error)
	CreateMCPToken(ctx context.Context, params db.CreateMCPTokenParams) (*db.MCPToken, error)
	ListMCPTokens(ctx context.Context, playerID uuid.UUID) ([]db.MCPToken, error)
	GetMCPToken(ctx context.Context, tokenID uuid.UUID) (*db.MCPToken, error)
	RevokeMCPToken(ctx context.Context, tokenID uuid.UUID) error
}

type pgMCPTokenStore struct {
	pool *pgxpool.Pool
}

func (s *pgMCPTokenStore) GenerateMCPToken() (raw, hash, prefix string, err error) {
	return db.GenerateMCPToken()
}

func (s *pgMCPTokenStore) CreateMCPToken(ctx context.Context, params db.CreateMCPTokenParams) (*db.MCPToken, error) {
	return db.CreateMCPToken(ctx, s.pool, params)
}

func (s *pgMCPTokenStore) ListMCPTokens(ctx context.Context, playerID uuid.UUID) ([]db.MCPToken, error) {
	return db.ListMCPTokens(ctx, s.pool, playerID)
}

func (s *pgMCPTokenStore) GetMCPToken(ctx context.Context, tokenID uuid.UUID) (*db.MCPToken, error) {
	return db.GetMCPToken(ctx, s.pool, tokenID)
}

func (s *pgMCPTokenStore) RevokeMCPToken(ctx context.Context, tokenID uuid.UUID) error {
	return db.RevokeMCPToken(ctx, s.pool, tokenID)
}

type MCPTokenHandler struct {
	Store MCPTokenStore
}

func NewMCPTokenHandler(pool *pgxpool.Pool) *MCPTokenHandler {
	return &MCPTokenHandler{Store: &pgMCPTokenStore{pool: pool}}
}

func (h *MCPTokenHandler) Routes() chi.Router {
	r := chi.NewRouter()
	r.Post("/", h.createToken)
	r.Get("/", h.listTokens)
	r.Delete("/{tokenID}", h.revokeToken)
	return r
}

func (h *MCPTokenHandler) createToken(w http.ResponseWriter, r *http.Request) {
	player := middleware.PlayerFromCtx(r.Context())
	if player == nil {
		renderError(w, apierror.Unauthorized())
		return
	}

	var body struct {
		Name      string  `json:"name"`
		ExpiresIn *string `json:"expires_in"`
	}
	if err := decodeJSON(r, &body); err != nil {
		renderError(w, err)
		return
	}
	if body.Name == "" {
		renderError(w, apierror.Unprocessable("name is required"))
		return
	}

	var expiresAt *time.Time
	if body.ExpiresIn != nil {
		now := time.Now().UTC()
		switch *body.ExpiresIn {
		case "24h":
			t := now.Add(24 * time.Hour)
			expiresAt = &t
		case "7d":
			t := now.Add(7 * 24 * time.Hour)
			expiresAt = &t
		default:
			renderError(w, apierror.Unprocessable("expires_in must be '24h', '7d', or null"))
			return
		}
	}

	raw, hash, prefix, err := h.Store.GenerateMCPToken()
	if err != nil {
		renderError(w, apierror.Internal("failed to generate token"))
		return
	}

	token, err := h.Store.CreateMCPToken(r.Context(), db.CreateMCPTokenParams{
		PlayerID:  player.ID,
		Name:      body.Name,
		TokenHash: hash,
		Prefix:    prefix,
		ExpiresAt: expiresAt,
	})
	if err != nil {
		renderError(w, apierror.Internal("failed to create token"))
		return
	}

	renderJSON(w, http.StatusCreated, map[string]any{
		"id":           token.ID,
		"name":         token.Name,
		"token":        raw,
		"token_prefix": token.TokenPrefix,
		"expires_at":   token.ExpiresAt,
		"created_at":   token.CreatedAt,
	})
}

func (h *MCPTokenHandler) listTokens(w http.ResponseWriter, r *http.Request) {
	player := middleware.PlayerFromCtx(r.Context())
	if player == nil {
		renderError(w, apierror.Unauthorized())
		return
	}

	tokens, err := h.Store.ListMCPTokens(r.Context(), player.ID)
	if err != nil {
		renderError(w, apierror.Internal("failed to list tokens"))
		return
	}

	now := time.Now().UTC()
	type tokenResp struct {
		ID          uuid.UUID  `json:"id"`
		Name        string     `json:"name"`
		TokenPrefix string     `json:"token_prefix"`
		ExpiresAt   *time.Time `json:"expires_at"`
		CreatedAt   time.Time  `json:"created_at"`
		LastUsedAt  *time.Time `json:"last_used_at"`
		IsExpired   bool       `json:"is_expired"`
	}

	resp := make([]tokenResp, len(tokens))
	for i, t := range tokens {
		resp[i] = tokenResp{
			ID:          t.ID,
			Name:        t.Name,
			TokenPrefix: t.TokenPrefix,
			ExpiresAt:   t.ExpiresAt,
			CreatedAt:   t.CreatedAt,
			LastUsedAt:  t.LastUsedAt,
			IsExpired:   t.ExpiresAt != nil && t.ExpiresAt.Before(now),
		}
	}
	renderJSON(w, http.StatusOK, resp)
}

func (h *MCPTokenHandler) revokeToken(w http.ResponseWriter, r *http.Request) {
	player := middleware.PlayerFromCtx(r.Context())
	if player == nil {
		renderError(w, apierror.Unauthorized())
		return
	}

	tokenIDStr := chi.URLParam(r, "tokenID")
	tokenID, err := uuid.Parse(tokenIDStr)
	if err != nil {
		renderError(w, apierror.NotFound("token not found"))
		return
	}

	token, err := h.Store.GetMCPToken(r.Context(), tokenID)
	if err != nil {
		renderError(w, apierror.NotFound("token not found"))
		return
	}
	if token.PlayerID != player.ID {
		renderError(w, apierror.Forbidden("cannot revoke another player's token"))
		return
	}

	if err := h.Store.RevokeMCPToken(r.Context(), tokenID); err != nil {
		renderError(w, apierror.Internal("failed to revoke token"))
		return
	}

	noContent(w)
}
